src/Pact/Native/Capabilities.hs

```diff
@@ -327,3 +336,61 @@ emitEventDef =
         DefcapManaged {} -> return ()
         DefcapEvent -> return ()
       _ -> evalError' i $ "emit-event: must be managed or event defcap"
+
+createPrincipalDef :: NativeDef
+createPrincipalDef =
+  defRNative "create-principal" createPrincipal'
+  (funType tTyString
+    [ ("guard", tTyGuard Nothing) ] )
+  [LitExample "(create-principal (read-keyset 'keyset))"]
+  "Create a principal which unambiguously identifies GUARD."
+  where
+    createPrincipal' :: RNativeFun e
+    createPrincipal' i [TGuard g _] =
+      toTerm <$> createPrincipal (getInfo i) g
+    createPrincipal' i as = argsError i as
+
+createPrincipal :: Info -> Guard (Term Name) -> Eval e Text
+createPrincipal i = \case
+  GPact (PactGuard pid n) -> do
+    chargeGas 1
+    pure $ "p:" <> asString pid <> ":" <> n
+  GKeySet (KeySet ks pf) -> case (toList ks,asString pf) of
+    ([k],"keys-all") -> do
+      chargeGas 1
+      pure $ "k:" <> asString k
+    (l,fun) -> do
+      a <- mkHash $ map _pubKey l
+      pure $ "w:" <> asString a <> ":" <> fun
+  GKeySetRef (KeySetName n) -> do
+    chargeGas 1
+    pure $ "r:" <> n
+  GModule (ModuleGuard mn n) -> do
+    chargeGas 1
+    pure $ "m:" <> asString mn <> ":" <> n
+  GUser (UserGuard uf args) -> do
+    a <- mkHash $ map toJSONPactValue args
+    pure $ "u:" <> asString uf <> ":" <> asString a
+  where
+    chargeGas amt = void $ computeGasCommit i "createPrincipal" (GPrincipal amt)
+    mkHash bss = do
+      let bs = mconcat bss
+      chargeGas $ 1 + (BS.length bs `div` 64) -- charge for 64 bytes of hashing
```

Comment:
Nitpick: could use `quot` since faster and length is never negative, but particularly I think 
```haskell
2 + (BS.length bs `quot` 64) 
```
Overcharges slightly in the case that `bs` is at least 1 digest in size, but is a bit more accurate. Don't think it's worth using quotRem for it.